src/libraries/System.Text.Json/tests/Serialization/ReferenceHandlerTests.cs

```diff
@@ -536,5 +532,133 @@ public static void ArrayPreserveDuplicateDictionaries()
             Assert.Same(rootCopy[0], rootCopy[1]);
         }
         #endregion
+
+        #region ReferenceResolver
+        [Fact]
+        public static void CustomReferenceResolver()
```

Comment:
Can we test scenarios where we have malicious behavior from custom resolvers e.g. when `alreadyExists` is always returned when calling `GetReference()` (`JsonException` cycle detected is expected here I believe)?